Analysis/Tasks/eventSelection.cxx

```diff
@@ -123,12 +133,97 @@ struct EventSelectionTask {
+    evsel(alias, bbT0A, bbT0C, bbV0A, bbV0C, bgV0A, bgV0C, bbZNA, bbZNC, bbFDA, bbFDC, bgFDA, bgFDC, foundFT0);
   }
 };
 
-WorkflowSpec defineDataProcessing(ConfigContext const&)
+// helper table and task for BCs<-->FT0s matching
+namespace o2::aod
 {
-  return WorkflowSpec{
-    adaptAnalysisTask<EventSelectionTask>("event-selection")};
+DECLARE_SOA_INDEX_TABLE_USER(SparseBCsFT0s, BCs, "MA_SP_BCS_FT0S", o2::aod::indices::BCId, o2::aod::indices::FT0Id);
 }
+
+struct EventSelectionTaskRun3Helper {
+  Builds<aod::SparseBCsFT0s> matchedBCsFT0s;
+  void process(aod::BCs, aod::FT0s)
+  {
+  }
+};
+
+struct EventSelectionTaskRun3 {
+  Produces<aod::EvSels> evsel;
+
+  EvSelParameters par;
+
+  void process(aod::Collision const& collision, aod::SparseBCsFT0s const& matchedBCsFT0s, aod::BCs const& bcs, aod::Zdcs const& zdcs, aod::FV0As const& fv0as, aod::FT0s const& ft0s, aod::FDDs const& fdds)
+  {
+    // calculating [minBC, maxBC] range for a collision
+    int deltaBC = std::ceil(collision.collisionTimeRes() / o2::constants::lhc::LHCBunchSpacingNS * 4);
+    int64_t colBC = collision.bc().globalBC();
+    int64_t maxColBC = colBC + deltaBC;
+    int64_t minColBC = colBC - deltaBC;
+
+    int64_t minFt0Dist = INT32_MAX;
+    int64_t foundFT0 = -1;
+    float timeA = -999.f;
+    float timeC = -999.f;
+
+    // calculating BCID iteration range
+    int64_t minGlobalBC = bcs.begin().globalBC();
+    int64_t maxGlobalBC = bcs.iteratorAt(bcs.size() - 1).globalBC();
+    int64_t startIt = minColBC - minGlobalBC;
+    int64_t endIt = maxColBC - minGlobalBC;
+
+    // iterating only through relevant BCIDs and
+    // searching for the nearest FT0
+    for (int64_t i = startIt; i <= endIt; i++) {
+      auto matched = matchedBCsFT0s.iteratorAt(i);
+      if (matched.has_ft0() == false) {
+        continue;
+      }
+      auto bc = bcs.iteratorAt(i);
+      int64_t ft0BC = bc.globalBC();
+      int64_t ft0Dist = abs(ft0BC - colBC);
+      if (ft0Dist < minFt0Dist) {
+        minFt0Dist = ft0Dist;
+        foundFT0 = matched.ft0().globalIndex();
+      }
+    }
+
+    if (minFt0Dist != INT32_MAX) {
+      auto ft0 = ft0s.iteratorAt(foundFT0);
+      timeA = ft0.timeA();
+      timeC = ft0.timeC();
+    }
+
+    bool bbZNA = 1;
+    bool bbZNC = 1;
+    bool bbV0A = 0;
+    bool bbV0C = 0;
+    bool bgV0A = 0;
+    bool bgV0C = 0;
+    bool bbFDA = 0;
+    bool bbFDC = 0;
+    bool bgFDA = 0;
+    bool bgFDC = 0;
+    bool bbT0A = timeA > par.fT0ABBlower && timeA < par.fT0ABBupper;
+    bool bbT0C = timeC > par.fT0CBBlower && timeC < par.fT0CBBupper;
+
+    int32_t alias[kNaliases] = {0};
+    // Fill event selection columns
+    // saving FT0 row index (foundFT0) for further analysis
+    evsel(alias, bbT0A, bbT0C, bbV0A, bbV0C, bgV0A, bgV0C, bbZNA, bbZNC, bbFDA, bbFDC, bgFDA, bgFDC, foundFT0);
+  }
+};
+
+WorkflowSpec defineDataProcessing(ConfigContext const& ctx)
+{
+  if (ctx.options().get<int>("selection-run") == 2) {
```

Comment:
@ekryshen @nburmaso The run 2 and run 3 case are separate analysis structs and chosen by a configurable? What speaks against putting them in separate executables instead?

Reply:
@jgrosseo I would consider this separation as a temporary solution. I hope we'll converge to a single task at some point

Reply:
@jgrosseo Note that this option is set to run2 by default so there is no issue with hyperloop for Run2-converted trains.

Reply:
@ekryshen I think the general logic of O2 is to split fine grained into workflows instead of using parameters. Common functions can always be used where the things intersect. Anyway, I think we don't agree here, so I merge this for now.

Reply:
@jgrosseo I hope we'll be able to merge it in one task without run2/run3 parameters. We just need to unify the process inputs:
process(aod::Collision const& collision, aod::SparseBCsFT0s const& matchedBCsFT0s, aod::BCs const& bcs, ...)
process(aod::Run2MatchedSparse::iterator const& collision, aod::BCsWithTimestamps const&, ... )

I believe it is doable but we might need Anton's expertise. In case we still need different branches in the process function for run2 and run3, we can always decide which branch to follow depending on run number.

Reply:
I'd also rather have two executables for the two different (Run2 vs Run3) configurations. That way the decision on what to run is made at the alihyperloop level, not user code.